pkg/agent/loadbalancer/servers.go

```diff
@@ -1,166 +1,421 @@
+func (sl *serverList) recordSuccess(srv *server, r reason) {
+	var new_state state
+	switch srv.state {
+	case stateFailed, stateUnchecked:
+		// dialed or health checked OK once, improve to recovering
+		new_state = stateRecovering
+	case stateRecovering:
+		if r == reasonHealthCheck {
+			// was recovering due to successful dial or first health check, can now improve
+			if len(srv.connections) > 0 {
+				// server accepted connections while recovering, attempt to go straight to active
+				new_state = stateActive
+			} else {
+				// no connections, just make it preferred
+				new_state = statePreferred
+			}
+		}
+	case stateHealthy:
+		if r == reasonDial {
+			// improve from healthy to active by being dialed
+			new_state = stateActive
+		}
+	case statePreferred:
+		if r == reasonDial {
+			// improve from healthy to active by being dialed
+			new_state = stateActive
+		} else {
+			if time.Now().Sub(srv.lastTransition) > time.Minute {
+				// has been preferred for a while without being dialed, demote to healthy
+				new_state = stateHealthy
 			}
 		}
 	}
 
-	lb.ServerAddresses = serverAddresses
-	lb.randomServers = append([]string{}, lb.ServerAddresses...)
-	rand.Shuffle(len(lb.randomServers), func(i, j int) {
-		lb.randomServers[i], lb.randomServers[j] = lb.randomServers[j], lb.randomServers[i]
-	})
-	// If the current server list does not contain the default server address,
-	// we want to include it in the random server list so that it can be tried if necessary.
-	// However, it should be treated as always failing health checks so that it is only
-	// used if all other endpoints are unavailable.
-	if !hasDefaultServer {
-		lb.randomServers = append(lb.randomServers, lb.defaultServerAddress)
-		if defaultServer, ok := lb.servers[lb.defaultServerAddress]; ok {
-			defaultServer.healthCheck = func() bool { return false }
-			lb.servers[lb.defaultServerAddress] = defaultServer
+	// no-op if state did not change
+	if new_state == stateInvalid {
+		return
+	}
+
+	// handle active transition and sort the server list while holding the lock
+	sl.mutex.Lock()
+	defer sl.mutex.Unlock()
+
+	// handle states of other servers when attempting to make this one active
+	if new_state == stateActive {
+		for _, s := range sl.servers {
+			if srv.address == s.address {
+				continue
+			}
+			switch s.state {
+			case stateFailed, stateStandby, stateRecovering, stateHealthy:
+				// close connections to other non-active servers whenever we have a new active server
+				defer s.closeAll()
+			case stateActive:
+				if len(s.connections) > len(srv.connections) {
+					// if there is a currently active server that has more connections that we do than we do,
+					// close our connections and go to preferred instead
+					new_state = statePreferred
+					defer srv.closeAll()
+				} else {
+					// otherwise, close its connections and demote it to preferred
+					s.state = statePreferred
+					defer s.closeAll()
```

Comment:
I'm confused by this use case. How would the agent have more connections with a server which was not the previously active one?

Reply:
The dialContext loop will dial all the servers in the list until one succeeds. Each dialing goroutine has its own copy of the list and may end up connected to a different server if they are stuck in recovering due to failing health checks. Once one passes health checks and becomes active, we should close all the connections to other servers so that they switch over to the active one with the most connections.